### DISCUSSION TITLE
How to add event handlers

### DISCUSSION THREAD

**Discussion by CharJon** [ds4dm/ecole#193] [Questions] [answered]
I am trying to add an event handler to an environment that keeps track of the dual bound history.
The handler gets added as part of the BranchingDynamics `reset_dynamics`. When executing the code I get

> Fatal Python error: PyThreadState_Get: no current thread
> Python runtime state: initialized
> 
> Thread 0x00007ff08fb9c740 (most recent call first):
>   File "notebooks/ecole_event_handler.py", line 121 in reset_dynamics
>   Aborted (core dumped)

You can find a slimmed down version of my code that produces the same error [here](https://gist.github.com/CharJon/b61e9e251d1abab76d2dac63bd55fd57).

I can not find an error in my approach, but maybe I am doing something wrong.

**Comment by AntoinePrv** [comment, marked as answer]
Hi @CharJon,

I think I understand what is going on. Python has a [Global Interpreter Lock](https://realpython.com/python-gil/) (GIL) that constraint it to be single threaded. When executing native (_e.g._ C/C++) code, this lock can be lifted to allow threading, which is what Ecole does.

PySCIPOpt however does not make this assumption. So when Ecole solves the `Model`, the _Python_ event handler eventually gets called but the GIL is never locked...

I have send a [PR](https://github.com/scipopt/PySCIPOpt/pull/505) to PySCIPOpt  to ensure the GIL is locked (you can already use it without it getting merged).
It should work, but our PySCIPOpt interface in Ecole still has a few rough edges... For instance, in your `DualBoundObservation`, you would now get 
```python
ReferenceError                            Traceback (most recent call last)
<ipython-input-2-367f052e7cd2> in eventinit(self)
     23 
     24     def eventinit(self):
---> 25         self.model.catchEvent(scip.SCIP_EVENTTYPE.LPEVENT, self)
     26 
     27     def eventexit(self):

ReferenceError: weakly-referenced object no longer exists
```
To avoid it, you need to keep a reference to the PySCIPOpt Model like so:
```python
    def before_reset(self, model):
        self.pysicpopt_model =  model.as_pyscipopt()
        self.pysicpopt_model.includeEventhdlr(self.dual_bound_handler, "Some handler", "Its description")
```
